data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -20,187 +8,41 @@
  */
 public class StructuredTraceGraph {
 
-  /* there could be multiple roots for partial trace and incomplete instrumented app spans */
-  private final Map<ByteBuffer, List<Event>> parentToChildrenEvents;
-  private final Map<ByteBuffer, Event> childToParentEvents;
-  /* entity have many-to-many relationship */
-  private final Map<String, List<Entity>> parentToChildrenEntities;
-  private final Map<String, List<Entity>> childToParentEntities;
+  private TraceEventsGraph traceEventsGraph;
+  private TraceEntitiesGraph traceEntitiesGraph;
 
-  /* these containers should be unmodifiable after initialization as we're exposing them via getters */
-  private Map<ByteBuffer, Event> eventMap;
-  private Map<String, Entity> entityMap;
-  private Set<Event> rootEvents;
-  private Set<Entity> rootEntities;
+  private static StructuredTraceGraph graph;
 
-  private StructuredTraceGraph() {
-    this.childToParentEntities = new HashMap<>();
-    this.parentToChildrenEntities = new HashMap<>();
-    this.childToParentEvents = new HashMap<>();
-    this.parentToChildrenEvents = new HashMap<>();
-  }
-
-  /**
-   * Create the instance with or without the full traversal graph includesFullGraph    Includes
-   * computing the full traversal graph or not
-   */
   public static StructuredTraceGraph createGraph(StructuredTrace trace) {
-    StructuredTraceGraph graph = new StructuredTraceGraph();
-    graph.buildEventMap(trace);
-    graph.buildEntityMap(trace);
-    graph.buildParentChildRelationship(trace);
-    graph.buildRootEvents(trace);
-    graph.buildRootEntities(trace);
-    return graph;
-  }
-
-  /**
-   * @return an immutable set containing the root events
-   */
-  public Set<Event> getRootEvents() {
-    return rootEvents;
-  }
-
-  /**
-   * @return an immutable set containing the root entities
-   */
-  public Set<Entity> getRootEntities() {
```

Comment:
As we are changing the interface, we will need to change the consumer, do we need StructureTraceGraph ? Or should we just keep the interface as is, and just fetch the data either from internal TraceEventGraph or TraceEntityGraph?

Reply:
What challenges do you see with changing the consumer?

Reply:
Now, `StructureTraceGraph` looks dummy holder for `TraceEventGraph` and `TraceEntityGraph`. So, I think, either we can remove it completely or we can maintain its existing public interface as-is for downstream consumption. So, all its public methods will internally be served via either `TraceEventGraph` or `TraceEntityGraph`. And both of them will become private in the context of `StructureTraceGraph`. And, we will only change in downstream `StructureTraceGraphBuiler` which will only build portion - either `TraceEventGraph` or `TraceEntityGraph` - on trace change.